hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -279,6 +297,87 @@ public BackgroundTaskResult call() throws Exception {
       return BackgroundTaskResult.EmptyTaskResult.newResult();
     }
 
+    private boolean checkSnapshotReclaimable(
+        Table<String, RepeatedOmKeyInfo> snapshotDeletedTable,
+        Table<String, OmKeyInfo> snapshotDeletedDirTable,
+        String snapshotBucketKey, String dbBucketKeyForDir) throws IOException {
+
+      boolean isDirTableCleanedUp = false;
+      boolean isKeyTableCleanedUp  = false;
+      try (TableIterator<String, ? extends Table.KeyValue<String,
+          RepeatedOmKeyInfo>> iterator = snapshotDeletedTable.iterator();) {
+        iterator.seek(snapshotBucketKey);
+        isKeyTableCleanedUp = iterator.hasNext() && iterator.next().getKey()
+            .startsWith(snapshotBucketKey);
+      }
+
+      try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+               iterator = snapshotDeletedDirTable.iterator()) {
+        iterator.seek(dbBucketKeyForDir);
+        isDirTableCleanedUp = iterator.hasNext() && iterator.next().getKey()
+            .startsWith(dbBucketKeyForDir);
+      }
+
+      return (isDirTableCleanedUp || snapshotDeletedDirTable.isEmpty()) &&
+          (isKeyTableCleanedUp || snapshotDeletedTable.isEmpty());
+    }
+
+    private long handleDirectoryCleanUp(
+        Table<String, OmKeyInfo> snapshotDeletedDirTable,
+        Table<String, OmDirectoryInfo> previousDirTable,
+        String dbBucketKeyForDir,
+        SnapshotInfo snapInfo, OmSnapshot omSnapshot,
+        List<String> dirsToMove) {
+
+      int dirNum = 0;
+      int subDirNum = 0;
+      int subFileNum = 0;
+      long remainNum = keyLimitPerSnapshot;
+      List<PurgePathRequest> purgePathRequestList = new ArrayList<>();
+      List<Pair<String, OmKeyInfo>> allSubDirList
+          = new ArrayList<>((int) remainNum);
```

Comment:
one less type conversion
```suggestion
      List<Pair<String, OmKeyInfo>> allSubDirList
          = new ArrayList<>(keyLimitPerSnapshot);
```